### PR TITLE
Add support for Qt 5.8 to Qt 5.15

### PR DESCRIPTION
**Pull request by Herrie82**
Various patches we carried in LuneOS while we upgraded through the various Qt 5.x versions. This has been tested with Qt 5.8 up to 5.15 and confirmed to be working.

Signed-off-by: Herman van Hazendonk <github.com@herrie.org>

### CODE REVIEW

**Comment by Herrie82** [comment]
This is in preparation for a possible Qt6 patch which we'll be sending upstream for review as well once it's ready.

File: customcontext/context.h [hunk 1]
```diff
@@ -59,22 +59,54 @@
...
 struct QSGMaterialType;
 #endif
 
+#if QT_VERSION >= 0x050800
+#define CONTEXT_CLASS Context
+#define CONTEXT_CLASS_BASE QSGDefaultContext
+#define RENDER_CONTEXT_CLASS RenderContext
+#define RENDER_CONTEXT_CLASS_BASE QSGDefaultRenderContext
+#define COMPILESHADER_METHOD compileShader
+#elif QT_VERSION >= 0x050200
```
Comment (pvuorela) [inline customcontext/context.h]: Don't think we need anymore support for below 5.2. By this point could just make this #else and drop the last set. Might allow to drop some definitions maybe?

Comment (Herrie82) [reply customcontext/context.h]: In our Qt 6 builds we've dropped it completely for now. The rework required was quite big and seems we can get away with not having the whole qtscenegraph-adaptation for now by the looks of it.

Comment (Herrie82) [reply customcontext/context.h]: > Don't think we need anymore support for below 5.2. By this point could just make this #else and drop the last set. Might allow to drop some definitions maybe?

Happy for suggestions there, just wanted to upstream some of our patches, so they would be useful for you guys down the line and we don't need to fork and keep carrying them :)

Comment (pvuorela) [reply customcontext/context.h]: Yea, I think we can have Qt 5.6 the oldest version that needs to be supported these days.

**Comment by Herrie82** [comment]
@pvuorela I removed the various Qt 5.2 checks and related bits with https://github.com/sailfishos/qtscenegraph-adaptation/pull/1/commits/c2926c615bbabd4927f47b3bbf3dfd7d9d157fc1 and cleaned up some deprecated functions in https://github.com/sailfishos/qtscenegraph-adaptation/pull/1/commits/7aa9f33bcf860016fd9dda626fad8e24db627a87

I've build tested this with Qt 5.15.2 here and can confirm it builds OK.

File: customcontext/texture/eglgralloctexture.cpp [hunk 2]
```diff
@@ -318,7 +318,7 @@ void EglGrallocTexture::bind()
 
         int error;
         while ((error = glGetError()) != GL_NO_ERROR)
-            qDebug() << "Error after glEGLImageTargetTexture2DOES" << hex << error;
+            qDebug() << "Error after glEGLImageTargetTexture2DOES" << Qt::hex << error;
```
Comment (pvuorela) [inline customcontext/texture/eglgralloctexture.cpp]: Qt::hex introduced in 5.14 so this needs some backward support.

File: customcontext/texture/eglgralloctexture.cpp [hunk 3]
```diff
@@ -148,7 +148,7 @@ NativeBuffer::NativeBuffer(const QImage &image)
     const int w = image.width();
     if (simpleCopy) {
         if (dbpl == image.bytesPerLine()) {
-            memcpy(data, image.constBits(), image.byteCount());
+            memcpy(data, image.constBits(), image.sizeInBytes());
```
Comment (pvuorela) [inline customcontext/texture/eglgralloctexture.cpp]: "This function was introduced in Qt 5.10."

File: customcontext/customcontext.pro [hunk 4]
```diff
@@ -3,10 +3,10 @@ TARGET=customcontext
 
 CONFIG += plugin
 
-contains(QT_VERSION, ^5\\.[2-9]\\..*) {
+equals(QT_MAJOR_VERSION, 5):greaterThan(QT_MINOR_VERSION, 1) {
```
Comment (pvuorela) [inline customcontext/customcontext.pro]: The whole customcontext_qt520 looks like something that can be removed now and the parts of the file with rules for < 5.2 qt.

File: customcontext/context.h [hunk 5]
```diff
@@ -59,34 +59,51 @@
...
+#if QT_VERSION >= 0x050800
+#include <private/qsgdefaultcontext_p.h>
+#include <private/qsgdefaultrendercontext_p.h>
+#endif
 
-#if QT_VERSION >= 0x050200
 struct QSGMaterialType;
+
+#define CONTEXT_CLASS Context
+#define RENDER_CONTEXT_CLASS RenderContext
```
Comment (pvuorela) [inline customcontext/context.h]: With <5.2 support removed, these don't have anymore variation. Could remove the definitions and instead just use target names in the code?

File: customcontext/context.cpp [hunk 6]
```diff
@@ -302,15 +262,25 @@ void Context::renderContextInitialized(QSGRenderContext *ctx)
...
-void CONTEXT_CLASS::initialize(QOpenGLContext *context)
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+void RENDER_CONTEXT_CLASS::initialize(const InitParams *inCtx)
 {
+#elif QT_VERSION >= 0x050800
+void RENDER_CONTEXT_CLASS::initialize(void *inCtx)
+{
+    QOpenGLContext *context = static_cast<QOpenGLContext*>(inCtx);
+#else
+void RENDER_CONTEXT_CLASS::initialize(QOpenGLContext *inCtx)
```
Comment (pvuorela) [inline customcontext/context.cpp]:
```
In file included from context.cpp:42:
context.h:70:30: error: no declaration matches ‘void CustomContext::RenderContext::initialize(QOpenGLContext*)’
 #define RENDER_CONTEXT_CLASS RenderContext
                              ^~~~~~~~~~~~~
context.cpp:280:6: note: in expansion of macro ‘RENDER_CONTEXT_CLASS’
 void RENDER_CONTEXT_CLASS::initialize(QOpenGLContext *inCtx)
      ^~~~~~~~~~~~~~~~~~~~
context.h:96:10: note: candidate is: ‘void CustomContext::RenderContext::initialize(void*)’
     void initialize(void *context);
          ^~~~~~~~~~
context.h:85:7: note: ‘class CustomContext::RenderContext’ defined here
 class RenderContext : public RENDER_CONTEXT_CLASS_BASE
       ^~~~~~~~~~~~~
make[1]: *** [Makefile:478: context.o] Error 1

```

**Review by pvuorela — COMMENTED** [review]
Looks better than earlier but still some gltiches to handle. Could also squash some commits to avoid changes back and forth.

**Comment by Herrie82** [comment]
@pvuorela I've pushed 2 commits now to address your comments, only tested on 5.15 here but seems to compile OK. Happy to receive feedback. Once you're happy with them, happy to quash them into a single commit.

File: customcontext/texture/eglgralloctexture.cpp (hunk 2 above)
Comment (Herrie82) [reply customcontext/texture/eglgralloctexture.cpp]: Fixed with https://github.com/sailfishos/qtscenegraph-adaptation/pull/1/commits/bb1d0094d4710788fe3c78f6a103f7d237c04967

File: customcontext/texture/eglgralloctexture.cpp (hunk 3 above)
Comment (Herrie82) [reply customcontext/texture/eglgralloctexture.cpp]: Fixed with https://github.com/sailfishos/qtscenegraph-adaptation/pull/1/commits/bb1d0094d4710788fe3c78f6a103f7d237c04967

File: customcontext/customcontext.pro (hunk 4 above)
Comment (Herrie82) [reply customcontext/customcontext.pro]: Fixed with https://github.com/sailfishos/qtscenegraph-adaptation/pull/1/commits/a70e9a9e883cd0e6e989e18462b33d356ad395ce

File: customcontext/context.h (hunk 5 above)
Comment (Herrie82) [reply customcontext/context.h]: Fixed with https://github.com/sailfishos/qtscenegraph-adaptation/pull/1/commits/a70e9a9e883cd0e6e989e18462b33d356ad395ce

File: customcontext/context.cpp (hunk 6 above)
Comment (Herrie82) [reply customcontext/context.cpp]: Probably fixed with https://github.com/sailfishos/qtscenegraph-adaptation/pull/1/commits/a70e9a9e883cd0e6e989e18462b33d356ad395ce but I only have 5.15 here to test

File: customcontext/context.cpp [hunk 7]
```diff
@@ -302,15 +262,25 @@ void Context::renderContextInitialized(QSGRenderContext *ctx)
...
-void CONTEXT_CLASS::initialize(QOpenGLContext *context)
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+void RenderContext::initialize(const InitParams *inCtx)
 {
+#elif QT_VERSION >= 0x050800
+void RenderContext::initialize(void *inCtx)
+{
+    QOpenGLContext *context = static_cast<QOpenGLContext*>(inCtx);
+#else
+void RenderContext::initialize(QOpenGLContext *inCtx)
```
Comment (pvuorela) [inline customcontext/context.cpp:280]: Unfortunately still not compiling:
```
context.cpp:280:6: error: no declaration matches ‘void CustomContext::RenderContext::initialize(QOpenGLContext*)’
 void RenderContext::initialize(QOpenGLContext *inCtx)
      ^~~~~~~~~~~~~
In file included from context.cpp:42:
context.h:93:10: note: candidate is: ‘void CustomContext::RenderContext::initialize(void*)’
     void initialize(void *context);
          ^~~~~~~~~~
context.h:82:7: note: ‘class CustomContext::RenderContext’ defined here
 class RenderContext : public RENDER_CONTEXT_CLASS_BASE
       ^~~~~~~~~~~~~
```

But now it's easier to see the problem. The method definition in the header seem reversed with >=5.8 and older vs the actual use here.

File: customcontext/context.cpp [hunk 8]
```diff
@@ -616,9 +572,13 @@ QQuickTextureFactory *Context::createTextureFactory(const QImage &image)
 
 
 
-void CONTEXT_CLASS::renderNextFrame(QSGRenderer *renderer, GLuint fbo)
+#if QT_VERSION >= 0x050800
+void RenderContext::renderNextFrame(QSGRenderer *renderer, uint fbo)
```
Comment (pvuorela) [inline customcontext/context.cpp:574]: This looks interesting too. So old version was with latter parameter QLuint, which is still happening below, but the header has also QLuint for >5.14 and uint version which didn't exist before. Or should it just be QLuint for everything?

**Review by pvuorela — COMMENTED** [review]
So could you still squash the commits and force push the changes. Bonus points for rebasing against current master.

File: customcontext/context.cpp (hunk 8 above)
Comment (pvuorela) [reply customcontext/context.cpp:574]: Looking at git version and history, no. GLuint was before 5.8 and from there on it's uint. So this is ok here, but the header as it wrong way. Also as 5.14 didn't have any changes there, could perhaps have it's own #if handling there.

**Comment by Herrie82** [comment]
I've updated the handling a bit. Hopefully it's more clear and works now for you as well? Compiled OK on 5.15 here

File: customcontext/context.h [hunk 9]
```diff
@@ -59,34 +59,53 @@
...
-#if QT_VERSION < 0x050600
-    QSGTexture *createTexture(const QImage &image) const;
-    QSGTexture *createTextureNoAtlas(const QImage &image) const;
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+    void initialize(const InitParams *inCtx);
+#elif (QT_VERSION >= QT_VERSION_CHECK(5, 8, 0) && QT_VERSION < QT_VERSION_CHECK(5, 14, 0))
+    void initialize(QOpenGLContext *context);
 #else
-    QSGTexture *createTexture(const QImage &image, uint flags) const;
+    void initialize(void *context);
```
Comment (pvuorela) [inline customcontext/context.h:96]: Sorry, this is still the wrong way. Cpp uses (QOpenGLContext *inCtx) for qt 5.6 and void* for 5.8 onwards.

File: customcontext/context.h [hunk 10]
```diff
@@ -59,34 +59,53 @@
...
+#else
     void renderNextFrame(QSGRenderer *renderer, GLuint fbo);
+#endif
 
-#if QT_VERSION < 0x050600
-    QSGTexture *createTexture(const QImage &image) const;
-    QSGTexture *createTextureNoAtlas(const QImage &image) const;
+#if QT_VERSION >= QT_VERSION_CHECK(5, 14, 0)
+    void initialize(const InitParams *inCtx);
+#elif (QT_VERSION >= QT_VERSION_CHECK(5, 8, 0) && QT_VERSION < QT_VERSION_CHECK(5, 14, 0))
```
Comment (pvuorela) [inline customcontext/context.h]: The #if above should already guarantee the version is below 5.14, and skipping that would make the rules similar between cpp and this.

**Review by pvuorela — COMMENTED** [review]
Almost there :)

File: customcontext/context.h (hunk 9 above)
Comment (pvuorela) [reply customcontext/context.h:96]: Asterisk got interpreted as styling by github, but you get the point.

Comment (Herrie82) [reply customcontext/context.h:96]: Like this? https://github.com/sailfishos/qtscenegraph-adaptation/pull/1/commits/f301bd1fb4c302562388c7218f0b8e6adb8882c4 I double checked the source code of 5.6, 5.8 and 5.15, should be OK now?

File: customcontext/context.h (hunk 10 above)
Comment (Herrie82) [reply customcontext/context.h]: OK removed it again

**Comment by Herrie82** [comment]
> rebas

It should cleanly merge. There was only this commit: https://github.com/webOS-ports/qtscenegraph-adaptation/compare/herrie/qt5.8+...sailfishos:qtscenegraph-adaptation:master

**Review by pvuorela — APPROVED** [review]
Now looking good. Nicely cleans up some old cruft too. Thanks!

**Comment by Herrie82** [comment]
@pvuorela Thnx for merge!

**Pull request merged**